kotlinx-coroutines-core/common/src/flow/StateFlow.kt

```diff
@@ -0,0 +1,411 @@
+ *     sumState.cancel() // stop computing sumFlow
+ * }
+ * ```
+ *
+ * **Preview feature**: The design of `stateIn` operator is tentative and could change.
+ *
+ * ### Strong equality-based conflation
+ *
+ * Values in state flow are conflated using [Any.equals] comparison in a similar way to
+ * [distinctUntilChanged] operator. It is used to conflate incoming updates
+ * to [value][MutableStateFlow.value] in [MutableStateFlow] and to suppress emission of the values to collectors
+ * when new value is equal to the previously emitted one. State flow behavior with classes that violate
+ * the contract for [Any.equals] is unspecified.
+ *
+ * Note, that this conflation logic is stronger that the conflation that is used by [conflate] operator,
+ * which is based on object identity. For example, consider this flow of strings that are emitted
+ * every 100ms:
+ *
+ * ```
+ * val flow = flowOf(0, 1, 0).map { it.toString() }.onEach { delay(100) }
+ * ```
+ *
+ * Conflating this flow before a slow collector (that takes 250 ms on each value), produces
+ * `[0, 0]`. The second zero string is not conflated, because it is a different [String] object with different
+ * identity:
+ *
+ * ```
+ * flow.conflate().onEach { delay(250) }.toList().let { println(it) }
+ * ```
+ *
+ * Converting the same flow to `StateFlow` and then feeding to slow collector, produces just `[0]`.
+ * The second zero string is conflated because it is equal to the previous zero:
+ *
+ * ```
+ * coroutineScope { // scope for stateIn
+ *     flow.stateIn(this).onEach { delay(250) }.toList().let { println(it) }
+ * }
+ * ```
+ *
+ * ### StateFlow vs ConflatedBroadcastChannel
+ *
+ * Conceptually state flow is similar to
+ * [ConflatedBroadcastChannel][kotlinx.coroutines.channels.ConflatedBroadcastChannel]
+ * and is designed to completely replace `ConflatedBroadcastChannel` in the future.
+ * It has the following important improvements:
+ *
+ * * `StateFlow` is simpler, because it does not have to implement all the [Channel] APIs, which allows
+ *   for faster, garbage-free implementation, unlike `ConflatedBroadcastChannel` implementation that
+ *   allocates objects on each emitted value.
+ * * `StateFlow` always has a value which can be safely read at any time via [value] property.
+ *    Unlike `ConflatedBroadcastChannel`, there is no way to create a state flow without a value.
+ * * `StateFlow` has a clear separation into a read-only `StateFlow` interface and a [MutableStateFlow].
+ * * `StateFlow` conflation works similarly to [Flow.conflate] operator in that the last emitted value
+ *    before close is received by collectors and is always available afterwards, unlike
+ *    `ConflatedBroadcastChannel` that does not keep the last value when closed.
+ * * `StateFlow` conflation is based on equality like [distinctUntilChanged] operator,
+ *    unlike conflation in `ConflatedBroadcastChannel` that is based on reference identity.
```

Comment:
This difference in how operators define equality seems arbitrary from the outside. Although even if you know that one uses a channel and one doesn't, it's still not obvious why the channel doesn't use `equals` (presumably performance?). I'm worried this inconsistency is going to be very confusing to people trying to learn this API for the first time, or even people who use it often (I am pretty sure I'm going to keep forgetting which operator does what). Especially coming from RxJava, where (I believe) all operators are consistent in this regard.

I don't think this should block this PR as it's a larger question, but is this something that could be addressed in the future?

Reply:
The rationale here is that `Channel` is a very low-level primitive, so it is Ok for it to be performance conscious, while `StateFlow` is a very high-level primitive, so it is Ok for it to rely on equality that might be broken, etc.

Reply:
Agreed Channel is low-level, but that behavior is exposed by `Flow.conflate`, which doesn't seem as low-level. Thinking about this more though, it probably doesn't matter too much since the whole purpose of the `conflate()` operator is to affect backpressure, and for that it doesn't matter how equality is checked. If you want distinct elements, then you should still use `distinctUntilChanged()`, even if it's right after a `conflate()`.